File: packages/contracts/src/starterKit.ts
import type { Dimensions, EntityId, IsoDateTime, Sha256 } from "./common.js";

export type StarterAssetKind =
  | "background" | "overlay" | "transition" | "font" | "music" | "sound-effect"
  | "presenter-style" | "presenter-portrait" | "lower-third" | "caption-style";

export interface StarterAssetLicense {
  status: "cleared" | "conditional" | "pending" | "unknown" | "restricted";
  expression: string;
  name: string;
  licenseUri?: string;
  copyrightNotice?: string;
  attributionRequired: boolean;
  attributionText?: string;
  redistributionAllowed: boolean;
  commercialUseAllowed: boolean;
  derivativesAllowed: boolean;
  exportAllowed: boolean;
  restrictions?: string[];
}

export interface StarterAssetProvenance {
  origin: "alystria-authored" | "third-party-open-source" | "system-font" | "generated" | "user-import" | "derived";
  creator: string;
  creationMethod: "procedural-code" | "hand-authored" | "recorded" | "font-distribution" | "generative-model" | "user-supplied" | "derived-edit";
  sourceUri?: string;
  sourceRevision?: string;
  tool?: string;
  model?: string;
  promptArtifactId?: EntityId;
  promptAvailability?: "artifact-recorded" | "conversation-retained" | "not-recorded" | "not-applicable";
  synthetic?: boolean;
  c2paStatus?: "present-embedded" | "present-sidecar" | "absent" | "invalid" | "unsupported" | "not-applicable";
  ingredientAssetIds?: EntityId[];
  recordedAt: IsoDateTime;
  reviewStatus: "verified" | "needs-review" | "quarantined";
  notes?: string;
}

export interface StarterAssetSource {
  delivery: "procedural" | "bundled-file" | "optional-download" | "installed-family" | "style-preset" | "placeholder";
  availability: "ready" | "optional" | "planned" | "unavailable";
  relativePath?: string;
  contentHash?: Sha256;
  byteSize?: number;
  downloadUri?: string;
  familyNames?: string[];
  fallbackAssetId?: EntityId;
}

export interface StarterAssetTechnicalSpec {
  mediaType: string;
  renderSafe: boolean;
  remoteFetchRequired: boolean;
  dimensions?: Dimensions;
  durationMs?: number;
  loopable?: boolean;
  sampleRate?: 44100 | 48000 | 96000;
  channels?: 1 | 2 | 6;
  integratedLufs?: number;
  fontWeights?: number[];
  variableFont?: boolean;
  scriptCoverage?: ("latin" | "latin-extended" | "devanagari" | "cyrillic" | "greek" | "arabic" | "cjk" | "math" | "symbols")[];
  rendererRecipe?: string;
  safeAreaPercent?: number;
  transparentBackground?: boolean;
}

export interface StarterAssetAccessibility {
  reducedMotionSafe: boolean;
  highContrastSafe: boolean;
  description?: string;
  transcriptLabel?: string;
}

export interface StarterAsset {
  id: EntityId;
  kind: StarterAssetKind;
  name: string;
  description: string;
  tags: string[];
  source: StarterAssetSource;
  license: StarterAssetLicense;
  provenance: StarterAssetProvenance;
  technical: StarterAssetTechnicalSpec;
  accessibility: StarterAssetAccessibility;
}

export interface StarterThemeDefaults {
  backgroundAssetId: EntityId;
  overlayAssetId?: EntityId;
  transitionAssetId: EntityId;
  displayFontAssetId: EntityId;
  bodyFontAssetId: EntityId;
  codeFontAssetId: EntityId;
  presenterStyleAssetId: EntityId;
  presenterPortraitAssetId?: EntityId;
  lowerThirdAssetId?: EntityId;
  captionStyleAssetId?: EntityId;
}

export interface StarterThemeAlternatives {
  backgroundAssetIds: EntityId[];
  transitionAssetIds: EntityId[];
  fontAssetIds: EntityId[];
  presenterStyleAssetIds: EntityId[];
  presenterPortraitAssetIds: EntityId[];
  musicAssetIds: EntityId[];
  soundEffectAssetIds: EntityId[];
}

export interface StarterThemePack {
  id: EntityId;
  themeId: "minimal" | "academic" | "modern-tech" | "notebook" | "documentary" | "playful" | "childrens-education" | "corporate-training" | "light" | "dark";
  name: string;
  description: string;
  defaults: StarterThemeDefaults;
  alternatives: StarterThemeAlternatives;
  audioDefaults: { musicEnabled: boolean; effectsEnabled: boolean; musicGainDb: number; effectsGainDb: number; duckingDb: number };
  presenterDefaults: { usage: "none" | "sparse" | "balanced" | "frequent"; placement: "left" | "right" | "alternating" | "contextual" | "full"; maximumCoveragePercent: number; allowUserPortrait: boolean };
}

export interface UserAssetSlot {
  id: EntityId;
  kind: "background" | "overlay" | "font" | "music" | "sound-effect" | "presenter-portrait" | "presenter-video" | "logo" | "lower-third" | "caption-style";
  name: string;
  description: string;
  acceptedMediaTypes: string[];
  acceptedExtensions?: string[];
  maximumBytes: number;
  multiple: boolean;
  rightsAttestationRequired: boolean;
  provenanceRequired: boolean;
  consentRequired: boolean;
  normalization: "image-srgb" | "video-rec709" | "audio-48khz" | "font-sanitized" | "data-only";
  minimumDimensions?: Dimensions;
  maximumDurationMs?: number;
  guidance?: string[];
}

export interface StarterKitManifest {
  schemaVersion: 1;
  id: EntityId;
  version: string;
  name: string;
  description: string;
  defaults: { musicEnabled: false; effectsEnabled: false; remoteFetchDuringRender: false; unknownRightsBlockExport: true };
  assets: StarterAsset[];
  themePacks: StarterThemePack[];
  userAssetSlots: UserAssetSlot[];
}

export interface StarterKitDiagnostic {
  code: string;
  severity: "warning" | "error" | "fatal";
  path: string;
  message: string;
}

/** Cross-record checks which JSON Schema cannot express. */
export function inspectStarterKit(manifest: StarterKitManifest): StarterKitDiagnostic[] {
  const diagnostics: StarterKitDiagnostic[] = [];
  const assetById = new Map<string, StarterAsset>();
  const seen = new Set<string>();
  const duplicate = (id: string, path: string, label: string): void => {
    if (seen.has(`${label}:${id}`)) diagnostics.push({ code: "starter.duplicate-id", severity: "fatal", path, message: `Duplicate ${label} id: ${id}` });
    seen.add(`${label}:${id}`);
  };
  manifest.assets.forEach((asset, index) => {
    duplicate(asset.id, `/assets/${index}/id`, "asset");
    assetById.set(asset.id, asset);
    if (asset.source.availability === "ready" && !asset.technical.renderSafe) diagnostics.push({ code: "starter.ready-asset-not-render-safe", severity: "fatal", path: `/assets/${index}/technical/renderSafe`, message: `Ready asset ${asset.id} is not render-safe` });
    if (asset.source.availability === "planned" && asset.license.exportAllowed) diagnostics.push({ code: "starter.placeholder-exportable", severity: "fatal", path: `/assets/${index}/license/exportAllowed`, message: `Placeholder ${asset.id} cannot be export-cleared before delivery and rights review` });
    if (asset.license.status !== "cleared" && asset.license.exportAllowed) diagnostics.push({ code: "starter.uncleared-export", severity: "fatal", path: `/assets/${index}/license`, message: `Uncleared asset ${asset.id} cannot permit export` });
    if (asset.technical.remoteFetchRequired) diagnostics.push({ code: "starter.remote-render-fetch", severity: "fatal", path: `/assets/${index}/technical/remoteFetchRequired`, message: `Starter asset ${asset.id} may not fetch remotely during render` });
  });
  const requireAsset = (id: string, expectedKind: StarterAssetKind | readonly StarterAssetKind[], path: string): void => {
    const asset = assetById.get(id);
    if (!asset) { diagnostics.push({ code: "starter.missing-asset", severity: "fatal", path, message: `Unknown starter asset: ${id}` }); return; }
    const expected = Array.isArray(expectedKind) ? expectedKind : [expectedKind];
    if (!expected.includes(asset.kind)) diagnostics.push({ code: "starter.wrong-asset-kind", severity: "error", path, message: `Asset ${id} has kind ${asset.kind}; expected ${expected.join(" or ")}` });
  };
  manifest.themePacks.forEach((pack, index) => {
    duplicate(pack.id, `/themePacks/${index}/id`, "theme pack");
    const base = `/themePacks/${index}`;
    requireAsset(pack.defaults.backgroundAssetId, ["background", "overlay"], `${base}/defaults/backgroundAssetId`);
    requireAsset(pack.defaults.transitionAssetId, "transition", `${base}/defaults/transitionAssetId`);
    requireAsset(pack.defaults.displayFontAssetId, "font", `${base}/defaults/displayFontAssetId`);
    requireAsset(pack.defaults.bodyFontAssetId, "font", `${base}/defaults/bodyFontAssetId`);
    requireAsset(pack.defaults.codeFontAssetId, "font", `${base}/defaults/codeFontAssetId`);
    requireAsset(pack.defaults.presenterStyleAssetId, "presenter-style", `${base}/defaults/presenterStyleAssetId`);
    if (pack.defaults.presenterPortraitAssetId) requireAsset(pack.defaults.presenterPortraitAssetId, "presenter-portrait", `${base}/defaults/presenterPortraitAssetId`);
    if (pack.defaults.overlayAssetId) requireAsset(pack.defaults.overlayAssetId, "overlay", `${base}/defaults/overlayAssetId`);
    if (pack.defaults.lowerThirdAssetId) requireAsset(pack.defaults.lowerThirdAssetId, "lower-third", `${base}/defaults/lowerThirdAssetId`);
    if (pack.defaults.captionStyleAssetId) requireAsset(pack.defaults.captionStyleAssetId, "caption-style", `${base}/defaults/captionStyleAssetId`);
    pack.alternatives.backgroundAssetIds.forEach((id, item) => requireAsset(id, ["background", "overlay"], `${base}/alternatives/backgroundAssetIds/${item}`));
    pack.alternatives.transitionAssetIds.forEach((id, item) => requireAsset(id, "transition", `${base}/alternatives/transitionAssetIds/${item}`));
    pack.alternatives.fontAssetIds.forEach((id, item) => requireAsset(id, "font", `${base}/alternatives/fontAssetIds/${item}`));
    pack.alternatives.presenterStyleAssetIds.forEach((id, item) => requireAsset(id, "presenter-style", `${base}/alternatives/presenterStyleAssetIds/${item}`));
    pack.alternatives.presenterPortraitAssetIds.forEach((id, item) => requireAsset(id, "presenter-portrait", `${base}/alternatives/presenterPortraitAssetIds/${item}`));
    pack.alternatives.musicAssetIds.forEach((id, item) => requireAsset(id, "music", `${base}/alternatives/musicAssetIds/${item}`));
    pack.alternatives.soundEffectAssetIds.forEach((id, item) => requireAsset(id, "sound-effect", `${base}/alternatives/soundEffectAssetIds/${item}`));
  });
  const requiredThemes = new Set(["minimal", "academic", "modern-tech", "notebook", "documentary", "playful", "childrens-education", "corporate-training", "light", "dark"]);
  manifest.themePacks.forEach((pack) => requiredThemes.delete(pack.themeId));
  for (const theme of requiredThemes) diagnostics.push({ code: "starter.missing-theme-pack", severity: "fatal", path: "/themePacks", message: `Missing starter pack for theme: ${theme}` });
  manifest.userAssetSlots.forEach((slot, index) => duplicate(slot.id, `/userAssetSlots/${index}/id`, "upload slot"));
  return diagnostics;
}

File: apps/desktop/src-tauri/src/starter_kit.rs
//! Typed desktop-side starter-kit trust boundary.
//!
//! JSON Schema is canonical; these closed serde records prevent a catalog from
//! smuggling executable fields into the privileged broker and add the relational
//! checks that schema alone cannot express.

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::{BTreeMap, BTreeSet};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterKitManifest {
    pub schema_version: u32,
    pub id: String,
    pub version: String,
    pub name: String,
    pub description: String,
    pub defaults: StarterKitDefaults,
    pub assets: Vec<StarterAsset>,
    pub theme_packs: Vec<StarterThemePack>,
    pub user_asset_slots: Vec<UserAssetSlot>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterKitDefaults {
    pub music_enabled: bool,
    pub effects_enabled: bool,
    pub remote_fetch_during_render: bool,
    pub unknown_rights_block_export: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterAsset {
    pub id: String,
    pub kind: String,
    pub name: String,
    pub description: String,
    pub tags: Vec<String>,
    pub source: StarterAssetSource,
    pub license: StarterAssetLicense,
    pub provenance: StarterAssetProvenance,
    pub technical: StarterAssetTechnical,
    pub accessibility: StarterAssetAccessibility,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterAssetSource {
    pub delivery: String,
    pub availability: String,
    pub relative_path: Option<String>,
    pub content_hash: Option<String>,
    pub byte_size: Option<u64>,
    pub download_uri: Option<String>,
    pub family_names: Option<Vec<String>>,
    pub fallback_asset_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterAssetLicense {
    pub status: String,
    pub expression: String,
    pub name: String,
    pub license_uri: Option<String>,
    pub copyright_notice: Option<String>,
    pub attribution_required: bool,
    pub attribution_text: Option<String>,
    pub redistribution_allowed: bool,
    pub commercial_use_allowed: bool,
    pub derivatives_allowed: bool,
    pub export_allowed: bool,
    pub restrictions: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterAssetProvenance {
    pub origin: String,
    pub creator: String,
    pub creation_method: String,
    pub source_uri: Option<String>,
    pub source_revision: Option<String>,
    pub tool: Option<String>,
    pub model: Option<String>,
    pub prompt_artifact_id: Option<String>,
    pub prompt_availability: Option<String>,
    pub synthetic: Option<bool>,
    pub c2pa_status: Option<String>,
    pub ingredient_asset_ids: Option<Vec<String>>,
    pub recorded_at: String,
    pub review_status: String,
    pub notes: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterAssetTechnical {
    pub media_type: String,
    pub render_safe: bool,
    pub remote_fetch_required: bool,
    pub dimensions: Option<Dimensions>,
    pub duration_ms: Option<u64>,
    pub loopable: Option<bool>,
    pub sample_rate: Option<u32>,
    pub channels: Option<u8>,
    pub integrated_lufs: Option<f64>,
    pub font_weights: Option<Vec<u16>>,
    pub variable_font: Option<bool>,
    pub script_coverage: Option<Vec<String>>,
    pub renderer_recipe: Option<String>,
    pub safe_area_percent: Option<f64>,
    pub transparent_background: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Dimensions {
    pub width: u32,
    pub height: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterAssetAccessibility {
    pub reduced_motion_safe: bool,
    pub high_contrast_safe: bool,
    pub description: Option<String>,
    pub transcript_label: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterThemePack {
    pub id: String,
    pub theme_id: String,
    pub name: String,
    pub description: String,
    pub defaults: StarterThemeDefaults,
    pub alternatives: StarterThemeAlternatives,
    pub audio_defaults: StarterAudioDefaults,
    pub presenter_defaults: StarterPresenterDefaults,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterThemeDefaults {
    pub background_asset_id: String,
    pub overlay_asset_id: Option<String>,
    pub transition_asset_id: String,
    pub display_font_asset_id: String,
    pub body_font_asset_id: String,
    pub code_font_asset_id: String,
    pub presenter_style_asset_id: String,
    pub presenter_portrait_asset_id: Option<String>,
    pub lower_third_asset_id: Option<String>,
    pub caption_style_asset_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterThemeAlternatives {
    pub background_asset_ids: Vec<String>,
    pub transition_asset_ids: Vec<String>,
    pub font_asset_ids: Vec<String>,
    pub presenter_style_asset_ids: Vec<String>,
    pub presenter_portrait_asset_ids: Vec<String>,
    pub music_asset_ids: Vec<String>,
    pub sound_effect_asset_ids: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterAudioDefaults {
    pub music_enabled: bool,
    pub effects_enabled: bool,
    pub music_gain_db: f64,
    pub effects_gain_db: f64,
    pub ducking_db: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StarterPresenterDefaults {
    pub usage: String,
    pub placement: String,
    pub maximum_coverage_percent: f64,
    pub allow_user_portrait: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct UserAssetSlot {
    pub id: String,
    pub kind: String,
    pub name: String,
    pub description: String,
    pub accepted_media_types: Vec<String>,
    pub accepted_extensions: Option<Vec<String>>,
    pub maximum_bytes: u64,
    pub multiple: bool,
    pub rights_attestation_required: bool,
    pub provenance_required: bool,
    pub consent_required: bool,
    pub normalization: String,
    pub minimum_dimensions: Option<Dimensions>,
    pub maximum_duration_ms: Option<u64>,
    pub guidance: Option<Vec<String>>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StarterKitValidationIssue {
    pub code: &'static str,
    pub path: String,
    pub message: String,
}

impl StarterKitManifest {
    pub fn from_value(value: Value) -> Result<Self, serde_json::Error> {
        serde_json::from_value(value)
    }

    pub fn validate(&self) -> Vec<StarterKitValidationIssue> {
        let mut issues = Vec::new();
        if self.schema_version != 1 {
            issue(
                &mut issues,
                "starter.unsupported-schema",
                "/schemaVersion",
                format!("Unsupported starter-kit schema: {}", self.schema_version),
            );
        }
        if self.defaults.music_enabled
            || self.defaults.effects_enabled
            || self.defaults.remote_fetch_during_render
            || !self.defaults.unknown_rights_block_export
        {
            issue(&mut issues, "starter.unsafe-defaults", "/defaults", "Starter-kit safety defaults must keep audio opt-in, disable render-time fetches, and block unknown rights".into());
        }

        let mut assets = BTreeMap::new();
        for (index, asset) in self.assets.iter().enumerate() {
            if assets.insert(asset.id.as_str(), asset).is_some() {
                issue(
                    &mut issues,
                    "starter.duplicate-id",
                    format!("/assets/{index}/id"),
                    format!("Duplicate asset id: {}", asset.id),
                );
            }
            if asset.source.availability == "ready" && !asset.technical.render_safe {
                issue(
                    &mut issues,
                    "starter.ready-asset-not-render-safe",
                    format!("/assets/{index}/technical/renderSafe"),
                    format!("Ready asset is not render-safe: {}", asset.id),
                );
            }
            if asset.technical.remote_fetch_required {
                issue(
                    &mut issues,
                    "starter.remote-render-fetch",
                    format!("/assets/{index}/technical/remoteFetchRequired"),
                    format!("Remote render fetch is forbidden: {}", asset.id),
                );
            }
            if asset.license.status != "cleared" && asset.license.export_allowed {
                issue(
                    &mut issues,
                    "starter.uncleared-export",
                    format!("/assets/{index}/license"),
                    format!("Uncleared asset cannot permit export: {}", asset.id),
                );
            }
            if matches!(
                asset.source.delivery.as_str(),
                "bundled-file" | "optional-download"
            ) && !asset
                .source
                .content_hash
                .as_deref()
                .is_some_and(valid_sha256)
            {
                issue(
                    &mut issues,
                    "starter.missing-integrity",
                    format!("/assets/{index}/source/contentHash"),
                    format!("File-backed asset lacks a SHA-256: {}", asset.id),
                );
            }
            if asset.source.delivery == "bundled-file" && asset.source.relative_path.is_none() {
                issue(
                    &mut issues,
                    "starter.missing-path",
                    format!("/assets/{index}/source/relativePath"),
                    format!("Bundled asset lacks a relative path: {}", asset.id),
                );
            }
        }

        let mut themes = BTreeSet::new();
        let mut pack_ids = BTreeSet::new();
        for (index, pack) in self.theme_packs.iter().enumerate() {
            if !pack_ids.insert(pack.id.as_str()) {
                issue(
                    &mut issues,
                    "starter.duplicate-id",
                    format!("/themePacks/{index}/id"),
                    format!("Duplicate theme pack id: {}", pack.id),
                );
            }
            themes.insert(pack.theme_id.as_str());
            for (field, id, kinds) in pack.references() {
                match assets.get(id) {
                    None => issue(
                        &mut issues,
                        "starter.missing-asset",
                        format!("/themePacks/{index}/{field}"),
                        format!("Unknown starter asset: {id}"),
                    ),
                    Some(asset) if !kinds.contains(&asset.kind.as_str()) => issue(
                        &mut issues,
                        "starter.wrong-asset-kind",
                        format!("/themePacks/{index}/{field}"),
                        format!("Asset {id} has kind {}", asset.kind),
                    ),
                    Some(_) => {}
                }
            }
        }
        for required in [
            "minimal",
            "academic",
            "modern-tech",
            "notebook",
            "documentary",
            "playful",
            "childrens-education",
            "corporate-training",
            "light",
            "dark",
        ] {
            if !themes.contains(required) {
                issue(
                    &mut issues,
                    "starter.missing-theme-pack",
                    "/themePacks",
                    format!("Missing starter pack for theme: {required}"),
                );
            }
        }
        let mut slots = BTreeSet::new();
        for (index, slot) in self.user_asset_slots.iter().enumerate() {
            if !slots.insert(slot.id.as_str()) {
                issue(
                    &mut issues,
                    "starter.duplicate-upload-slot",
                    format!("/userAssetSlots/{index}/id"),
                    format!("Duplicate upload slot: {}", slot.id),
                );
            }
        }
        issues
    }
}

type Reference<'a> = (&'static str, &'a str, &'static [&'static str]);

impl StarterThemePack {
    fn references(&self) -> Vec<Reference<'_>> {
        let mut values = vec![
            (
                "defaults/backgroundAssetId",
                self.defaults.background_asset_id.as_str(),
                &["background", "overlay"][..],
            ),
            (
                "defaults/transitionAssetId",
                self.defaults.transition_asset_id.as_str(),
                &["transition"][..],
            ),
            (
                "defaults/displayFontAssetId",
                self.defaults.display_font_asset_id.as_str(),
                &["font"][..],
            ),
            (
                "defaults/bodyFontAssetId",
                self.defaults.body_font_asset_id.as_str(),
                &["font"][..],
            ),
            (
                "defaults/codeFontAssetId",
                self.defaults.code_font_asset_id.as_str(),
                &["font"][..],
            ),
            (
                "defaults/presenterStyleAssetId",
                self.defaults.presenter_style_asset_id.as_str(),
                &["presenter-style"][..],
            ),
        ];
        optional_ref(
            &mut values,
            "defaults/overlayAssetId",
            self.defaults.overlay_asset_id.as_deref(),
            &["overlay"],
        );
        optional_ref(
            &mut values,
            "defaults/presenterPortraitAssetId",
            self.defaults.presenter_portrait_asset_id.as_deref(),
            &["presenter-portrait"],
        );
        optional_ref(
            &mut values,
            "defaults/lowerThirdAssetId",
            self.defaults.lower_third_asset_id.as_deref(),
            &["lower-third"],
        );
        optional_ref(
            &mut values,
            "defaults/captionStyleAssetId",
            self.defaults.caption_style_asset_id.as_deref(),
            &["caption-style"],
        );
        list_refs(
            &mut values,
            "alternatives/backgroundAssetIds",
            &self.alternatives.background_asset_ids,
            &["background", "overlay"],
        );
        list_refs(
            &mut values,
            "alternatives/transitionAssetIds",
            &self.alternatives.transition_asset_ids,
            &["transition"],
        );
        list_refs(
            &mut values,
            "alternatives/fontAssetIds",
            &self.alternatives.font_asset_ids,
            &["font"],
        );
        list_refs(
            &mut values,
            "alternatives/presenterStyleAssetIds",
            &self.alternatives.presenter_style_asset_ids,
            &["presenter-style"],
        );
        list_refs(
            &mut values,
            "alternatives/presenterPortraitAssetIds",
            &self.alternatives.presenter_portrait_asset_ids,
            &["presenter-portrait"],
        );
        list_refs(
            &mut values,
            "alternatives/musicAssetIds",
            &self.alternatives.music_asset_ids,
            &["music"],
        );
        list_refs(
            &mut values,
            "alternatives/soundEffectAssetIds",
            &self.alternatives.sound_effect_asset_ids,
            &["sound-effect"],
        );
        values
    }
}

fn optional_ref<'a>(
    values: &mut Vec<Reference<'a>>,
    field: &'static str,
    id: Option<&'a str>,
    kinds: &'static [&'static str],
) {
    if let Some(id) = id {
        values.push((field, id, kinds));
    }
}

fn list_refs<'a>(
    values: &mut Vec<Reference<'a>>,
    field: &'static str,
    ids: &'a [String],
    kinds: &'static [&'static str],
) {
    values.extend(ids.iter().map(|id| (field, id.as_str(), kinds)));
}

fn valid_sha256(value: &str) -> bool {
    value.len() == 64
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

fn issue(
    issues: &mut Vec<StarterKitValidationIssue>,
    code: &'static str,
    path: impl Into<String>,
    message: String,
) {
    issues.push(StarterKitValidationIssue {
        code,
        path: path.into(),
        message,
    });
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_unknown_fields_at_the_deserialization_boundary() {
        let result = StarterKitManifest::from_value(serde_json::json!({
            "schemaVersion": 1, "id": "starter.test", "version": "1.0.0", "name": "Test", "description": "Test",
            "defaults": { "musicEnabled": false, "effectsEnabled": false, "remoteFetchDuringRender": false, "unknownRightsBlockExport": true },
            "assets": [], "themePacks": [], "userAssetSlots": [], "postInstall": "run-me"
        }));
        assert!(result.is_err());
    }

    #[test]
    fn sha256_check_requires_lowercase_hex() {
        assert!(valid_sha256(&"a".repeat(64)));
        assert!(!valid_sha256(&"G".repeat(64)));
        assert!(!valid_sha256("abc"));
    }
}
